homeassistant/components/zerproc/light.py

```diff
@@ -0,0 +1,194 @@
+"""Zerproc light platform."""
+import asyncio
+from datetime import timedelta
+import logging
+from typing import Callable, List, Optional
+
+import pyzerproc
+
+from homeassistant.components.light import (
+    ATTR_BRIGHTNESS,
+    ATTR_HS_COLOR,
+    SUPPORT_BRIGHTNESS,
+    SUPPORT_COLOR,
+    Light,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import EVENT_HOMEASSISTANT_STOP
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.event import async_track_time_interval
+from homeassistant.helpers.typing import HomeAssistantType
+import homeassistant.util.color as color_util
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+SUPPORT_ZERPROC = SUPPORT_BRIGHTNESS | SUPPORT_COLOR
+
+DISCOVERY_INTERVAL = timedelta(seconds=60)
+
+
+def connect_light(light: pyzerproc.Light) -> Optional[pyzerproc.Light]:
+    """Attempt to connect to the light, and return the entity."""
+    try:
+        light.connect(auto_reconnect=True)
+    except pyzerproc.ZerprocException:
+        _LOGGER.debug("Unable to connect to '%s'", light.address, exc_info=True)
+        return None
+
+    return light
+
+
+async def discover_entities(
+    hass: HomeAssistant, async_add_entities: Callable[[List[Entity], bool], None]
+) -> List[Entity]:
+    """Attempt to discover new lights."""
+    lights = await hass.async_add_executor_job(pyzerproc.discover)
+
+    # Filter out already discovered lights
+    new_lights = [
+        light
+        for light in lights
+        if light.address not in hass.data[DOMAIN]["light_entities"]
+    ]
+
+    entities = []
+    for light in await asyncio.gather(
+        *[hass.async_add_executor_job(connect_light, light) for light in new_lights]
```

Comment:
Instead of spawning a lot of executor jobs, do it all in 1 job. We only have limited amount of threads. 

I would probably suggest to convert this whole method to be `sync` since it's actually just a bunch of sync methods except for the last line.